### ISSUE TITLE
anonymous inner classes aren't recognized in exclude list

### ISSUE THREAD

**Issue by bshannon** [spotbugs/spotbugs#498]
(I accidentally filed this in the [discuss subproject](https://github.com/spotbugs/discuss/issues/22) the first time.)

I'm converting from FindBugs to SpotBugs. I have an exclude list with this entry:

    <Match>
        <Or>
            <Class name="com.sun.mail.imap.DefaultFolder$1"/>
            <Class name="com.sun.mail.imap.DefaultFolder$2"/>
            <Class name="com.sun.mail.imap.IMAPFolder$1"/>
            <Class name="com.sun.mail.imap.IMAPFolder$4"/>
            <Class name="com.sun.mail.imap.IMAPFolder$7"/>
            <Class name="com.sun.mail.imap.IMAPFolder$13"/>
            <Class name="com.sun.mail.imap.IMAPFolder$20"/>
            <Class name="com.sun.mail.imap.IdleManager$2"/>
        </Or>
        <Bug pattern="SIC_INNER_SHOULD_BE_STATIC_ANON"/>
    </Match>

FindBugs ignores these errors. SpotBugs does not.

Changing the entry to use a regular expression instead of an exact class name does work.

(I'm calling FindBugs/SpotBugs from Maven, if that matters.)

**Comment by KengoTODA** [comment]
Are you using Java 8 or 9?

In my understanding SpotBugs introduced [no change for filters](https://github.com/spotbugs/spotbugs/blob/3.1.0/CHANGELOG.md), so I just guess that this problem comes from other place such as JDK.

* https://github.com/spotbugs/spotbugs/commits/3.1.0/spotbugs/src/main/java/edu/umd/cs/findbugs/filter
* https://github.com/spotbugs/spotbugs/commits/9751a22ac595fbb0e2989357a361de7346d3eaa9/findbugs/src/main/java/edu/umd/cs/findbugs/filter

**Comment by bshannon** [comment]
JDK 8.  All I did was change FindBugs to SpotBugs.  Everything else is the same.

**Comment by iloveeclipse** [comment]
Which new warnings are found now? Can it be that SpotBugs simply can find warnings not found by FindBugs? We fixed lot of issues.

**Comment by bshannon** [comment]
No.  They both find the same errors.  The problem is with the ability to EXCLUDE those errors.

**Comment by sdavids** [comment]
I stumbled upon this issue by migrating to [JUnit 5 Nested Tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-nested):
```java
package test;

import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

class SpotbugsTest {

  @Nested
  class Inner {

    @Test
    void test() {
      assertTrue(true);
    }
  }
}
```

```
Gradle 4.10.2
JVM 1.8.0_181 (Oracle Corporation 25.181-b13)
```

```groovy
plugins {
  id 'com.github.spotbugs' version '1.6.3'
}

spotbugs {
  toolVersion = '3.1.7'
  excludeFilter = project.file('gradle/conf/spotbugs-exclude-filter.xml')
}
```

```xml
<FindBugsFilter
    xmlns="https://github.com/spotbugs/filter/3.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="https://github.com/spotbugs/filter/3.0.0 https://raw.githubusercontent.com/spotbugs/spotbugs/release-3.1/spotbugs/etc/findbugsfilter.xsd">
  <Match>
    <Class name="test.OuterTest$Inner"/>
    <Bug pattern="SIC_INNER_SHOULD_BE_STATIC"/>
  </Match>
</FindBugsFilter>
```

```
> Task :spotbugsTest
M P SIC: Should OuterTest$Inner be a _static_ inner class?  At SpotbugsTest.java:[lines 10-15]
```

**Comment by ThrawnCA** [comment]
Is there a specific version of SpotBugs that introduced this behavior? Does it occur with version 3.1.0?

**Comment by sdavids** [comment]
It does occur with 3.1.0.

**Comment by sdavids** [comment]
My [issue](https://github.com/spotbugs/spotbugs/issues/498#issuecomment-423849720) does not occur with Gradle SpotBugs plugin 1.6.9 and SpotBugs 3.1.11.
